Source/Diagnostics/ReducedDiags/ParticleExtrema.cpp

```diff
@@ -0,0 +1,393 @@
+            return std::sqrt(1.0 + us*inv_c2);
+        });
+        ParallelDescriptor::ReduceRealMax(gmax);
+
+        // wmin
+        Real wmin = ReduceMin( myspc,
+        [=] AMREX_GPU_HOST_DEVICE (const PType& p) -> Real
+        { return p.rdata(PIdx::w); });
+        ParallelDescriptor::ReduceRealMin(wmin);
+
+        // wmax
+        Real wmax = ReduceMax( myspc,
+        [=] AMREX_GPU_HOST_DEVICE (const PType& p) -> Real
+        { return p.rdata(PIdx::w); });
+        ParallelDescriptor::ReduceRealMax(wmax);
+
+#if (defined WARPX_QED)
+        // compute chimin and chimax
+        Real chimin = 0.0_rt;
+        Real chimax = 0.0_rt;
+        if (myspc.has_breit_wheeler() || myspc.has_quantum_sync())
+        {
+            chimin = ReduceMin( myspc,
+            [=] AMREX_GPU_HOST_DEVICE (const PType& p) -> Real
+            {
+#if (defined WARPX_DIM_RZ)
+                ParticleReal xp = p.pos(0)*std::cos(p.rdata(PIdx::theta));
+                ParticleReal yp = p.pos(0)*std::sin(p.rdata(PIdx::theta));
+#elif (defined WARPX_DIM_XZ)
+                ParticleReal xp = p.pos(0);
+                ParticleReal yp = 0.0_rt;
+#else
+                ParticleReal xp = p.pos(0);
+                ParticleReal yp = p.pos(1);
+#endif
+                ParticleReal zp = p.pos(index_z);
+                ParticleReal ex = 0._rt, ey = 0._rt, ez = 0._rt;
+                ParticleReal bx = 0._rt, by = 0._rt, bz = 0._rt;
+
+                Array4<const amrex::Real> ex_arr;
+                Array4<const amrex::Real> ey_arr;
+                Array4<const amrex::Real> ez_arr;
+                Array4<const amrex::Real> bx_arr;
+                Array4<const amrex::Real> by_arr;
+                Array4<const amrex::Real> bz_arr;
+                IndexType ex_type, ey_type, ez_type;
+                IndexType bx_type, by_type, bz_type;
+                GpuArray<amrex::Real, 3> dx_arr, xyzmin_arr;
+                Dim3 lo;
+
+                int n_rz_azimuthal_modes;
+                int nox;
+                bool galerkin_interpolation;
+
+                // gather E and B
+                doGatherShapeN(xp, yp, zp, ex, ey, ez, bx, by, bz,
+                    ex_arr, ey_arr, ez_arr, bx_arr, by_arr, bz_arr,
+                    ex_type, ey_type, ez_type,
+                    bx_type, by_type, bz_type,
+                    dx_arr, xyzmin_arr, lo,
+                    n_rz_azimuthal_modes, nox, galerkin_interpolation);
+
+                // compute chi
+                Real px = p.rdata(PIdx::ux)*m;
+                Real py = p.rdata(PIdx::uy)*m;
+                Real pz = p.rdata(PIdx::uz)*m;
+                if ( myspc.AmIA<PhysicalSpecies::photon>() )
+                { Real chi = chi_photon(px, py, pz, ex, ey, ez, bx, by, bz); }
+                else
+                { Real chi = chi_lepton(px, py, pz, ex, ey, ez, bx, by, bz); }
+
+                return 0.0_rt;
+            });
+            ParallelDescriptor::ReduceRealMin(chimin);
+        }
+#endif
```

Comment:
@atmyers Hi Andrew, does this part of the code look correct?

Reply:
Great! 
I just have a couple of comments:
- I agree with @atmyers that the external fields should be taken into account. To this end I guess that you should generate these two functors `auto get_externalE = GetExternalEField  (pti, offset);` and `auto get_externalB = GetExternalBField  (pti, offset);` just before the for loop. Then, before calling `doGatherShapeN`, you can do:
```
                    ParticleReal ex = 0._rt, ey = 0._rt, ez = 0._rt;
                    get_externalE(i, ex, ey, ez);
                    ParticleReal bx = 0._rt, by = 0._rt, bz = 0._rt;
                    get_externalB(i, bx, by, bz);
```  
- Shouldn't the functor inside `ReduceMin` return `chi` ?